### PR TITLE
Python3 bindings and get virtual page API

### PR DESCRIPTION
**Pull request by masthoon**
Hi, 

This PR adds Python 3 bindings (tested on Python 3.6 64 bits ; require some modification for 3.8) and some new API:
- GetDirectoryTableBase
- TranslateVirtualToPhysical
- GetVirtualPage

*WIP: Python38, auto tests of Python bindings, appveyor+travis build.*

### CODE REVIEW

File: src/lib/kdmp-parser.cc [hunk 1]
```diff
@@ -324,3 +345,96 @@ KernelDumpParser::GetPhysicalPage(const uint64_t PhysicalAddress) const {
 
   return Pair->second;
 }
+
+const uint64_t 
+KernelDumpParser::TranslateVirtualToPhysical(const uint64_t VirtualAddress, const uint64_t DirectoryTableBase) const {
+
+  //
+  // If DirectoryTableBase not provided uses the one from the dump header and remove PCID bits
```
Comment (0vercl0k) [inline src/lib/kdmp-parser.cc:353]: is not provided?

File: src/lib/kdmp-parser.cc [hunk 2]
```diff
@@ -324,3 +345,96 @@ KernelDumpParser::GetPhysicalPage(const uint64_t PhysicalAddress) const {
...
+
+  const uint64_t PteGpa = PtBase + GuestAddress.PtIndex * 8;
+  const MMPTE_HARDWARE Pte(UtilRead8(PteGpa));
+  if (!Pte.Present) {
+    return 0;
+  }
+
+  const uint64_t PageBase = Pte.PageFrameNumber * 0x1000;
+  return PageBase + GuestAddress.Offset;
+
```
Comment (0vercl0k) [inline src/lib/kdmp-parser.cc:418]: nits: extra-line

File: src/python/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# Mastho - 2020
+link_libraries(kdmp-parser)
+
+include_directories("C:\\Python36\\include")
```
Comment (0vercl0k) [inline src/python/CMakeLists.txt:4]: might need to do that differently as it needs to also compile on Linux - there might be a magic cmake module that finds those paths?

File: src/lib/kdmp-parser.cc [hunk 4]
```diff
@@ -217,6 +217,23 @@ bool KernelDumpParser::BuildPhysmemFullDump() {
   return true;
 }
 
+const uint64_t KernelDumpParser::UtilRead8(const uint64_t PhysicalAddress) const {
```
Comment (0vercl0k) [inline src/lib/kdmp-parser.cc:220]: How about VirtRead8 instead?

File: src/lib/kdmp-parser.cc (hunk 1 above)
Comment (0vercl0k) [reply src/lib/kdmp-parser.cc:353]: what PCID bits btw?

File: src/lib/kdmp-parser.cc [hunk 5]
```diff
@@ -324,3 +345,96 @@ KernelDumpParser::GetPhysicalPage(const uint64_t PhysicalAddress) const {
 
   return Pair->second;
 }
+
+const uint64_t 
+KernelDumpParser::TranslateVirtualToPhysical(const uint64_t VirtualAddress, const uint64_t DirectoryTableBase) const {
```
Comment (0vercl0k) [inline src/lib/kdmp-parser.cc:350]: what about VirtTranslate?

File: src/python/python-kdmp.h [hunk 6]
```diff
@@ -0,0 +1,121 @@
+#define PY_SSIZE_T_CLEAN
+
+#include <Python.h>
+#include "kdmp-parser.h"
+#define PAGE_SIZE 0x1000
```
Comment (0vercl0k) [inline src/python/python-kdmp.h:5]: Can use Page::Size?

File: src/python/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+link_directories("C:\\Python36\\libs")
```
Comment (0vercl0k) [inline src/python/CMakeLists.txt:5]: i guess it also breaks on x64 as the path needs to be adapted

Comment (0vercl0k) [reply src/python/CMakeLists.txt:5]: https://cmake.org/cmake/help/v3.12/module/FindPython3.html#module:FindPython3 this looks like the right thing to use

Comment (0vercl0k) [reply src/python/CMakeLists.txt:5]: example: https://github.com/JonathanSalwan/Triton/blob/master/CMakeLists.txt#L45

File: src/lib/kdmp-parser.cc (hunk 4 above)
Comment (masthoon) [reply src/lib/kdmp-parser.cc:220]: PhyRead8?

Comment (0vercl0k) [reply src/lib/kdmp-parser.cc:220]: Oops yes, sorry :-D

**Comment by masthoon** [comment]
**YO**
Ok it has been tested for py36/37/38 on Windows but the AppVeyor and Travis integration use Python 3.6 to make the script easier. Small python test have been integrated to *--run-tests* as well (*Python tests: All good!*).

No issue with AppVeyor build.
Travis doesn't seem to have 32 bits Python installed by default, so the py3 bindings won't compile on X86 build:
**Python 3.6 development library not found**. So I didn't implement the tests on it [builder.py](https://github.com/0vercl0k/kdmp-parser/pull/3/commits/66c5ee54f4559669c9062b693f6545c06f0c11e6#diff-83aeab46a47575c8469d5d040f4eab0cR30)

File: src/python/python-kdmp.cc [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+    Self->DumpParser = nullptr;
+
+    //
+    // Parse Python argument (expect a string i.e. the file path of the dump)
+    //    *PyErr_Format returns nullptr to raise the exception*
+    //
+
+    char *DumpPath = nullptr;
+    if (!PyArg_ParseTuple(Args,"s", &DumpPath)) {
+        return PyErr_Format(PyExc_TypeError, "dump() expected a string");
```
[the remark this answers is not in the document]
Comment (masthoon) [reply src/python/python-kdmp.cc:23]: Type->tp_free(Self); in destruct

[the remark this answers is not in the document]
Comment (masthoon) [reply src/python/python-kdmp.cc:23]: oh indeed

File: builder.py [hunk 9]
```diff
@@ -90,7 +109,26 @@ def test(arch, configuration, dmp_path):
...
+
+    if ret != 0: return ret
+
+    # Linux build move build/.../libkdmp.so to bin/.../kdmp.so
+    linux_mod = os.path.join('build', dir_name, 'src', 'python', 'libkdmp.so')
+    if is_linux and os.path.exists(linux_mod):
+        os.rename(linux_mod, os.path.join(bin_dir, 'kdmp.so'))
+    
+    # Python3 bindings test if kdmp module was created
+    if any('kdmp' in f for f in os.listdir(bin_dir)):
```
[the remark this answers is not in the document]
Comment (masthoon) [reply builder.py:122]: kdmp.pyd kdmp_d.pyd kdmp.so

[the remark this answers is not in the document]
Comment (0vercl0k) [reply builder.py:122]: I see - why do we care about kdmp*_d* btw?

File: src/lib/kdmp-parser.cc [hunk 10]
```diff
@@ -324,3 +345,95 @@ KernelDumpParser::GetPhysicalPage(const uint64_t PhysicalAddress) const {
 
   return Pair->second;
 }
+
+const uint64_t 
+KernelDumpParser::VirtTranslate(const uint64_t VirtualAddress, const uint64_t DirectoryTableBase) const {
+
+  //
+  // If DirectoryTableBase is null ; use the one from the dump header and clear PCID bits (bits 11:0)
```
[the remark this answers is not in the document]
Comment (masthoon) [reply src/lib/kdmp-parser.cc:353]: Page::Align does the job

**Comment by masthoon** [comment]
Most comments fixed.
Only missing 32 bits Python tests on Linux (how to install py32 and debug libs?).

File: src/lib/kdmp-parser.cc [hunk 11]
```diff
@@ -324,3 +345,96 @@ KernelDumpParser::GetPhysicalPage(const uint64_t PhysicalAddress) const {
...
+
+  const uint64_t PdBase = Pdpte.PageFrameNumber * Page::Size;
+  if (Pdpte.LargePage) {
+    return (PdBase & 0xffff'ffff'c000'0000) + (VirtualAddress & 0x3fff'ffff);
+  }
+
+  const uint64_t PdeGpa = PdBase + GuestAddress.PdIndex * 8;
+  const MMPTE_HARDWARE Pde(PhyRead8(PdeGpa));
+  if (!Pde.Present) {
+    return 0;
```
Comment (0vercl0k) [inline src/lib/kdmp-parser.cc:397]: Shouldn't there be other messages here and below? They are error cases as well

File: src/python/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,27 @@
+# Mastho - 2020
+link_libraries(kdmp-parser)
+find_package (Python3 COMPONENTS Interpreter Development)
+
+# Python on Windows Debug build only load $_d.pyd and require python_d.exe
+if(CMAKE_BUILD_TYPE STREQUAL "Debug" AND WIN32)
+    set(PYMODULE kdmp_d CACHE INTERNAL "")
```
Comment (0vercl0k) [inline src/python/CMakeLists.txt:7]: What's the `CACHE INTERNAL ""` part for?

File: src/python/python-kdmp.cc [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+    // Initialize python.
+    //
+
+    Py_Initialize();
+
+    //
+    // Register PythonDumpParserType.
+    //
+
+	if(PyType_Ready(&PythonDumpParserType) < 0)
```
Comment (0vercl0k) [inline src/python/python-kdmp.cc:243]: The indentation doesn't look right here - I can run clang-format once merged if you want me to

File: src/tests/tests.py [hunk 14]
```diff
@@ -19,9 +19,24 @@ def test(bin_dir, dmp_path):
...
+        dmp_path
+    )
+
+    print('Launching "{0}"..'.format(' '.join(cmd)))
+    return subprocess.call(cmd)
+
 def main():
     parser = argparse.ArgumentParser('Run test')
     parser.add_argument('--bindir', required = True)
+    parser.add_argument('--pymodule', required = True, nargs = '?')
```
Comment (0vercl0k) [inline src/tests/tests.py:39]: Can't you just set the argument as `required=False`? I think it'd have the behavior you want

File: .github/workflows/kdmp-parser.yml [hunk 15]
```diff
@@ -13,8 +13,10 @@ jobs:
 
     - name: Install dependencies
       run: |
+        sudo dpkg --add-architecture i386
```
Comment (0vercl0k) [inline .github/workflows/kdmp-parser.yml:16]: Let's forget about Linux 32 bits - I've tried a few things but I am not sure how to tell `CMake` / `FindPython` to look for a different endianess..

File: .github/workflows/kdmp-parser.yml [hunk 16]
```diff
@@ -13,8 +13,10 @@ jobs:
...
         sudo apt-get -y update
         sudo apt install -y g++-multilib ninja-build
+        sudo apt install -y libpython3-dev:i386
```
Comment (0vercl0k) [inline .github/workflows/kdmp-parser.yml:19]: Same here

File: src/python/CMakeLists.txt (hunk 12 above)
Comment (masthoon) [reply src/python/CMakeLists.txt:7]: PYMODULE is used in another cmake script so need CACHE keyword. INTERNAL probably not required.

File: src/python/python-kdmp.cc (hunk 13 above)
Comment (masthoon) [reply src/python/python-kdmp.cc:243]: yes please,  I will resolve the last comments and rebase the commit first.

File: src/tests/tests.py (hunk 14 above)
Comment (masthoon) [reply src/tests/tests.py:39]: I tried and it doesn't work because --pymodule argument is always present (see cmake file) but the value may be empty.

**Comment by masthoon** [comment]
So I squashed all the commits into 1.
I added GetBugCheckParameters API to retrieve the bugcheck code and parameters.

Regarding the tests: 
- No tests for GetBugCheckParameters _(it works locally :D)_
- Linux 64 bits release and debug **PASSED**
- Linux 32 bits **SKIPPED** because no 32 bits python executable found
- Windows 32/64 release **PASSED**
- Windows 32/64 debug **SKIPPED** because debug libs not installed (https://github.com/actions/setup-python/issues/86) _(it works locally too :D)_

Please run clang-format if someday you accept the PR.
Let me know if you need more changes.

**Comment by 0vercl0k** [comment]
🎉

**Pull request merged**